app/dao/notifications_dao.py

```diff
@@ -807,14 +807,19 @@ def get_service_ids_with_notifications_on_date(notification_type, date):
     start_date = get_london_midnight_in_utc(date)
     end_date = get_london_midnight_in_utc(date + timedelta(days=1))
 
-    return {
-        row.service_id
-        for row in db.session.query(
-            Notification.service_id
+    subqueries = union(*[
+        db.session.query(
+            table.service_id.label('service_id')
         ).filter(
-            Notification.notification_type == notification_type,
+            table.notification_type == notification_type,
             # using >= + < is much more efficient than date(created_at)
-            Notification.created_at >= start_date,
-            Notification.created_at < end_date,
-        ).distinct()
+            table.created_at >= start_date,
+            table.created_at < end_date,
+        )
+        for table in [Notification, NotificationHistory]
```

Comment:
the where clause in all of these has been: `where notification_type = 'sms' and created_at > '2022-05-02 23:00' and created_at < '2022-05-03 23:00'`

----------
so as a benchmark, here's the original query
```sql
-- original query
select distinct service_id from notifications where ...
```
https://explain.dalibo.com/plan# execution time 3m40s

---------
so i first tried running your query as proposed in the PR
```sql
-- this PR's query
select distinct * from 
(
    select service_id from notifications where ...
    union
    select service_id from notification_history where ...
)
```
https://explain.dalibo.com/plan/ynG - 39m19s (!)

----------
then i thought it might be merging two giant lists so tried nesting the `distinct` calls. (in retrospect this is really simple for a computer to do and i probably didn't need to spend the time doing this)
```sql
-- with inner distincts as well
select distinct * from
(
    select distinct service_id from notifications where ...
    union
    select distinct service_id from notification_history where ...
)
```
https://explain.dalibo.com/plan/p1Q - 43m (a bit longer but i also did some stuff in the background during the explain which might have impacted our available read IOPS)

----------

i also just looked at the execution plans and noticed they're both sequential scans on both tables. Even the regular notifications table one is relatively slow! We've got an index that includes key type. 
```sql
-- looking at key_type separately
select distinct service_id from notifications where ... and key_type = 'team'
union 
select distinct service_id from notifications where ... and key_type = 'normal'
```
https://explain.dalibo.com/plan/o3L - 7m50s - didn't help (just duplicated the work). i forget how indexes work.

--------

then i realised we could join ft_notification_status rather than history
```sql
-- looking at ft_notification_status
 select distinct * from (
    select service_id from notifications where notification_type = 'sms' and created_at > '2022-05-02 23:00' and created_at < '2022-05-03 23:00'
    union
    select service_id from ft_notification_status where ft_notification_status.notification_type= 'sms' and bst_date = '2022-05-03'
)
```
https://explain.dalibo.com/plan/Ohn 1m56s! i think this might include the db starting to cache the notifications table result. But the ft_notification_status table takes 2ms to hit.

it potentially has really niche risks of not picking up services if all the nightly tasks go wrong for multiple days. but i think it's a fair assumption that we'll either have notifications in notification table OR have already processed this on a previous evening's billing task.

Reply:
This is a great idea 💯. Thanks for (a) questioning the performance; and (b) taking the time to investigate it. I've implemented your last suggestion and confirmed the query plan - around 2 mins for the 10th of May.

I was initially surprised the performance is so bad for the history table. I was thinking of [similar but different work on queries to it](https://github.com/alphagov/notifications-api/pull/3417). I did a little further investigation and it seems that, even with a further restriction on key_type, Postgres still believes the index isn't selective enough to compensate for a full table scan. Looking at the index stats, I think I can see why - the relevant index is half the size of the table, so comparable to the real thing.

```
 public | notification_history      | table | xxx | 1157 GB | 
```
```
 public | ix_notification_history_service_id_composite | index | xxx | notification_history | 456 GB | 
```

Anyway, could you take another look?

(I'm open to doing a more succinct `.union()` - I just find the `union(...)` version a bit more readable.)